proxy/src/main/java/com/wavefront/agent/preprocessor/predicate/ContainsPredicate.java

```diff
@@ -0,0 +1,38 @@
+package com.wavefront.agent.preprocessor.predicate;
+
+import com.wavefront.agent.preprocessor.PreprocessorUtil;
+
+import java.util.List;
+
+import wavefront.report.ReportPoint;
+import wavefront.report.Span;
+
+/**
+ * Predicate mimicking {@link String#contains(java.lang.CharSequence)} for Wavefront reportable entities.
+ *
+ * @author Anil Kodali (akodali@vmware.com).
+ */
+public class ContainsPredicate<T> extends ComparisonPredicate<T>{
+
+  public ContainsPredicate(String scope, String value) {
+    super(scope, value);
+  }
+
+  @Override
+  public boolean test(T t) {
+    if (t instanceof ReportPoint) {
```

Comment:
While this approach would work, using raw types and checking `instanceof` in a critical path (especially considering that the number of times test(T) is called is going to be many times more than the incoming data rate) is less than ideal in our case. Sometimes it's difficult to work around it and you are stuck with dealing with generics wildcards, but in this case it should be fairly easy - just implement `ReportPointContainsPredicate extends ComparisonPredicate<ReportPoint>` and `SpanContainsPredicate extends ComparisonPredicate<Span>` separately. As you are building your predicate tree you are already aware of the rule context, whether it's ReportPoint or Span, so that should be easier. While it does look like more code, it will be a more performant and scalable solution, and it also gives you flexibility to implement things like allContains/allStartWith/etc that would be applicable only to spans (or any possible future entity where annotations are a multimap).

Reply:
done, Thanks for the detailed review comments.